pkg/frontend/nvme.go

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2023 Intel Corporation
+
+// Package frontend implements the FrontEnd APIs (host facing) of the storage Server
+package frontend
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+
+	pb "github.com/opiproject/opi-api/storage/v1alpha1/gen/go"
+	"github.com/opiproject/opi-spdk-bridge/pkg/models"
+	"github.com/ulule/deepcopier"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+const (
+	mevTransportType = "npi"
+)
+
+// CreateNVMeController creates an NVMe Controller
+func (s *Server) CreateNVMeController(ctx context.Context, in *pb.CreateNVMeControllerRequest) (*pb.NVMeController, error) {
+	log.Printf("CreateNVMeController: Received from client: %v", in)
+
+	err := verifyCreateNvmeControllerRequestArgs(in)
+	if err != nil {
+		return nil, err
+	}
+
+	subsys, ok := s.nvme.Subsystems[in.NvMeController.Spec.SubsystemId.Value]
+	if !ok {
+		msg := fmt.Sprintf("unable to find subsystem with id %v", in.NvMeController.Spec.SubsystemId.Value)
+		log.Printf("error: %v", msg)
+		return nil, status.Error(codes.FailedPrecondition, msg)
+	}
+	log.Printf("Found subsystem: %v", subsys.Spec.Nqn)
+
+	controllerID := in.NvMeController.Spec.Id.Value
+	_, ok = s.nvme.Controllers[controllerID]
+	if ok {
+		msg := fmt.Sprintf("Already existing controller with id %v", controllerID)
+		log.Printf("error: %v", msg)
```

Comment:
for idempotency, why instead of error, not to return the old controller here ?

Reply:
opi-spdk-bridge needs to be reworked in terms of idempotency first, let's do that in a further step here as well